test/EndToEndTests/Tests/Core/Microsoft.OData.Core.E2E.Tests/WriteJsonPayloadTests/WritePayloadTests.cs

```diff
@@ -0,0 +1,837 @@
+            {
+                orderFeed.SetSerializationInfo(new ODataResourceSerializationInfo() { NavigationSourceName = "Orders", NavigationSourceEntityTypeName = NameSpacePrefix + "Order" });
+            }
+
+            await odataWriter.WriteStartAsync(orderFeed);
+
+            var orderEntry1 = TestsHelper.CreateOrderEntry1(hasModel);
+            await odataWriter.WriteStartAsync(orderEntry1);
+
+            var orderEntry1Navigation1 = new ODataNestedResourceInfo()
+            {
+                Name = "Customer",
+                IsCollection = false,
+                Url = new Uri(_baseUri + "Orders(-10)/Customer")
+            };
+
+            await odataWriter.WriteStartAsync(orderEntry1Navigation1);
+            await odataWriter.WriteEndAsync();
+
+            var orderEntry1Navigation2 = new ODataNestedResourceInfo()
+            {
+                Name = "Login",
+                IsCollection = false,
+                Url = new Uri(_baseUri + "Orders(-10)/Login")
+            };
+
+            await odataWriter.WriteStartAsync(orderEntry1Navigation2);
+            await odataWriter.WriteEndAsync();
+
+            // Finish writing orderEntry1.
+            await odataWriter.WriteEndAsync();
+
+            var orderEntry2Wrapper = TestsHelper.CreateOrderEntry2(hasModel);
+
+            var orderEntry2Navigation1 = new ODataNestedResourceInfo()
+            {
+                Name = "Customer",
+                IsCollection = false,
+                Url = new Uri(_baseUri + "Orders(-9)/Customer")
+            };
+
+            var orderEntry2Navigation2 = new ODataNestedResourceInfo()
+            {
+                Name = "Login",
+                IsCollection = false,
+                Url = new Uri(_baseUri + "Orders(-9)/Login")
+            };
+
+            orderEntry2Wrapper.NestedResourceInfoWrappers = orderEntry2Wrapper.NestedResourceInfoWrappers.Concat(
+                new[]
+                {
+                    new ODataNestedResourceInfoWrapper() { NestedResourceInfo = orderEntry1Navigation1 },
+                    new ODataNestedResourceInfoWrapper() { NestedResourceInfo = orderEntry2Navigation2 }
+                });
+
+            await ODataWriterHelper.WriteResourceAsync(odataWriter, orderEntry2Wrapper);
+
+            // Finish writing the feed.
+            await odataWriter.WriteEndAsync();
+
+            Stream stream = await responseMessage.GetStreamAsync();
+
+            return await TestsHelper.ReadStreamContentAsync(stream);
+        }
+
+        private async Task<string> WriteAnExpandedCustomerEntryAsync(
+            TestStreamResponseMessage responseMessage,
+            ODataWriter odataWriter,
+            bool hasModel,
+            string mimeType)
+        {
+            ODataResourceWrapper customerEntry = TestsHelper.CreateCustomerEntry(hasModel);
+
+            var loginFeed = new ODataResourceSet() { Id = new Uri(_baseUri + "Customers(-9)/Logins") };
+            if (!hasModel)

```

Comment:
The reliance of compex helper code in our tests is still a concern. I will not block PR because of this but I'd want us to brainstorm and think of ways to avoid this pattern in our code. The risk with reliance on such helpers is that when reading the tests these helpers are blackboxes that we may not fully understand, and we may not always inspect exhaustively. Our tests relies on these helpers being correct, but we don't have guarantees that they are correct since we don't have test cases for most helpers. So if we have a bug in the helpers, it could cause a faulty test to pass and hide a bug.

Reply:
I'll create a work item to track this.